File: bundler/src/main/java/com/hendraanggrian/bundler/Bundler.java
package com.hendraanggrian.bundler;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Fragment;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.Log;

import org.json.JSONObject;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public final class Bundler {

    private static final String TAG = "Bundler";
    private static boolean DEBUG;
    @Nullable private static Map<String, Constructor<? extends BundleBinding>> BINDINGS;

    private Bundler() {
    }

    public static void setDebug(boolean debug) {
        Bundler.DEBUG = debug;
    }

    /**
     * Bind extra fields in support Fragment with default behavior.
     */
    public static void bindExtras(@NonNull android.support.v4.app.Fragment target) {
        bindExtras(target, target.getArguments());
    }

    /**
     * Bind extra fields in Fragment with default behavior,
     * not available pre-11.
     */
    @RequiresApi(11)
    @TargetApi(11)
    public static void bindExtras(@NonNull Fragment target) {
        bindExtras(target, target.getArguments());
    }

    /**
     * Bind extra fields in Activity with default behavior.
     */
    public static void bindExtras(@NonNull Activity target) {
        bindExtras(target, target.getIntent());
    }

    /**
     * Bind extra fields in target from source Bundle attached to Intent.
     */
    public static <T> void bindExtras(@NonNull T target, @NonNull Intent source) {
        bindExtras(target, source.getExtras());
    }

    /**
     * Bind extra fields in target from source Bundle.
     */
    public static <T> void bindExtras(@NonNull T target, @NonNull Bundle source) {
        BundleBinding binding = createBinding(
                target.getClass(),
                BindExtra.SUFFIX,
                new Class<?>[]{target.getClass(), Bundle.class},
                new Object[]{target, source}
        );
        if (DEBUG) printContent(binding.source);
    }

    /**
     * Bind state fields in target from source Bundle.
     */
    public static <T> void bindStates(@NonNull T target, @NonNull Bundle source) {
        BundleBinding binding = createBinding(
                target.getClass(),
                BindState.SUFFIX,
                new Class<?>[]{target.getClass(), Bundle.class},
                new Object[]{target, source}
        );
        if (DEBUG) printContent(binding.source);
    }

    @NonNull
    public static Bundle wrapExtras(@NonNull Class<?> targetClass, @NonNull Object arg) {
        return wrapExtras(targetClass, new ArrayList<>(Collections.singletonList(arg)));
    }

    @NonNull
    public static Bundle wrapExtras(@NonNull Class<?> targetClass, @NonNull Object... args) {
        return wrapExtras(targetClass, new ArrayList<>(Arrays.asList(args)));
    }

    @NonNull
    public static Bundle wrapExtras(@NonNull Class<?> targetClass, @NonNull List<?> args) {
        BundleBinding binding = createBinding(
                targetClass,
                BindExtra.SUFFIX,
                new Class<?>[]{List.class},
                new Object[]{args}
        );
        if (DEBUG) printContent(binding.source);
        return binding.source;
    }

    @NonNull
    public static Bundle onSaveInstanceState(@NonNull Bundle source, @NonNull Class<?> targetClass, @NonNull List<?> args) {
        BundleBinding binding = createBinding(
                targetClass,
                BindState.SUFFIX,
                new Class<?>[]{List.class},
                new Object[]{args}
        );
        source.putAll(binding.source);
        if (DEBUG) printContent(binding.source);
        return source;
    }

    @NonNull
    @SuppressWarnings("TryWithIdenticalCatches")
    private static BundleBinding createBinding(
            @NonNull Class<?> targetClass,
            @NonNull String targetClassSuffix,
            @NonNull Class<?>[] constructorParameterTypes,
            @NonNull Object[] constructorParameters
    ) {
        if (DEBUG) Log.d(TAG, "Looking up constructor for " + targetClass.getName());
        Constructor<? extends BundleBinding> constructor = findBinding(targetClass, targetClassSuffix, constructorParameterTypes);
        if (constructor == null) {
            if (DEBUG)
                Log.d(TAG, "Ignored because no constructor was found in " + targetClass.getSimpleName());
            return BundleBinding.EMPTY;
        }
        try {
            return constructor.newInstance(constructorParameters);
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Unable to invoke " + constructor, e);
        } catch (InstantiationException e) {
            throw new RuntimeException("Unable to invoke " + constructor, e);
        } catch (InvocationTargetException e) {
            Throwable cause = e.getCause();
            if (cause instanceof RuntimeException)
                throw (RuntimeException) cause;
            if (cause instanceof Error)
                throw (Error) cause;
            throw new RuntimeException("Unable to create constructor instance.", cause);
        }
    }

    @Nullable
    @SuppressWarnings("unchecked")
    private static Constructor<? extends BundleBinding> findBinding(
            @NonNull Class<?> targetClass,
            @NonNull String targetClassSuffix,
            @NonNull Class<?>[] constructorParameterTypes
    ) {
        if (BINDINGS == null) {
            BINDINGS = new WeakHashMap<>();
        }
        String bindingKey = targetClass.toString() + constructorParameterTypes.length;
        Constructor<? extends BundleBinding> binding = BINDINGS.get(bindingKey);
        if (binding != null) {
            if (DEBUG) Log.d(TAG, "HIT: Cache found in binding weak map.");
            return binding;
        }
        String targetClassName = targetClass.getName();
        if (targetClassName.startsWith("android.") || targetClassName.startsWith("java.")) {
            if (DEBUG) Log.d(TAG, "MISS: Reached framework class. Abandoning search.");
            return null;
        }
        try {
            binding = (Constructor<? extends BundleBinding>) targetClass
                    .getClassLoader()
                    .loadClass(targetClassName + targetClassSuffix)
                    .getConstructor(constructorParameterTypes);
            if (DEBUG) Log.d(TAG, "HIT: Loaded binding class, caching in weak map.");
        } catch (ClassNotFoundException e) {
            if (DEBUG)
                Log.d(TAG, "Not found. Trying superclass " + targetClass.getSuperclass().getName());
            Class<?> targetSuperclass = targetClass.getSuperclass();
            //region abstract classes binding fix
            if (constructorParameterTypes[0] == targetClass && constructorParameterTypes[0].getSuperclass() == targetSuperclass) {
                constructorParameterTypes[0] = targetSuperclass;
            }
            //endregion
            binding = findBinding(targetSuperclass, targetClassSuffix, constructorParameterTypes);
        } catch (NoSuchMethodException e) {
            throw new RuntimeException("Unable to find binding constructor for " + targetClassName, e);
        }
        BINDINGS.put(bindingKey, binding);
        return binding;
    }

    private static void printContent(@NonNull Bundle bundle) {
        try {
            JSONObject jsonObject = new JSONObject();
            for (String key : bundle.keySet()) {
                jsonObject.put(key, bundle.get(key));
            }
            String json = jsonObject.toString(4);
            for (String line : json.split("\\r?\\n")) {
                Log.d(TAG, line);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: bundler/src/com/hendraanggrian/bundler/Bundler.kt
package com.hendraanggrian.bundler

import android.content.Intent
import android.os.Bundle
import android.util.Log
import com.hendraanggrian.bundler.internal.BundleBinding
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException
import java.util.WeakHashMap
import org.json.JSONObject

object Bundler {
    private const val TAG = "Bundler"
    private var BINDINGS: MutableMap<String, Constructor<BundleBinding>>? = null
    private var DEBUG = false

    /** When set to true, will print Bundler operation in debug level. */
    fun setDebug(debug: Boolean) {
        DEBUG = debug
    }

    internal fun debug(message: String) {
        if (DEBUG) {
            Log.d(TAG, message)
        }
    }

    internal fun debug(message: Bundle) {
        if (DEBUG) {
            try {
                val jsonObject = JSONObject()
                for (key in message.keySet()) {
                    jsonObject.put(key, message[key])
                }
                val json = jsonObject.toString(4)
                val split = mutableListOf(*json.split("\\r?\\n").toTypedArray())
                val lastIndex = split.size - 1
                if (split[lastIndex].isEmpty()) {
                    split.removeAt(lastIndex)
                }
                split.forEach(::debug)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    /**
     * Bind fields annotated with [BindExtra] from source [Intent].
     * @param source intent to extract.
     * @param target fields' owner.
     * @throws RuntimeException when constructor of binding class cannot be found.
     */
    fun bindExtras(source: Intent, target: Any) =
        bindExtras(source.extras ?: error("No extras found in this Intent."), target)

    /**
     * Bind fields annotated with [BindExtra] from source [Bundle].
     * @param source bundle to extract.
     * @param target fields' owner.
     * @throws RuntimeException when constructor of binding class cannot be found.
     */
    fun bindExtras(source: Bundle, target: Any) {
        val binding = createBinding(
            target.javaClass, BindExtra.SUFFIX,
            arrayOf(Bundle::class.java, target.javaClass),
            arrayOf(source, target)
        )
        debug(binding.source)
    }

    fun wrapExtras(targetCls: Class<*>, args: List<*>): Bundle {
        val binding = createBinding(
            targetCls, BindExtra.SUFFIX,
            arrayOf(MutableList::class.java),
            arrayOf(args)
        )
        val extras = binding.source
        debug(extras)
        return extras
    }

    fun restoreStates(source: Bundle, target: Any) {
        val binding = createBinding(
            target.javaClass, BindState.SUFFIX,
            arrayOf(Bundle::class.java, target.javaClass),
            arrayOf(source, target)
        )
        debug(binding.source)
    }

    fun saveStates(source: Bundle, target: Any): Bundle {
        val binding = createBinding(
            target.javaClass, BindState.SUFFIX,
            arrayOf(target.javaClass, Bundle::class.java),
            arrayOf(target, source)
        )
        val states = binding.source
        source.putAll(states)
        debug(states)
        return source
    }

    private fun createBinding(cls: Class<*>, clsSuffix: String, paramTypes: Array<Class<*>>, params: Array<*>):
        BundleBinding {
        val clsName = cls.simpleName
        debug("Looking up constructor for $clsName")
        val constructor = findBindingConstructor(cls, clsSuffix, paramTypes)
        if (constructor == null) {
            debug("Ignored because no constructor was found in $clsName")
            return BundleBinding.EMPTY
        }
        return try {
            constructor.newInstance(*params)
        } catch (e: IllegalAccessException) {
            throw RuntimeException("Unable to invoke $constructor", e)
        } catch (e: InstantiationException) {
            throw RuntimeException("Unable to invoke $constructor", e)
        } catch (e: InvocationTargetException) {
            when (val cause = e.cause) {
                is RuntimeException -> throw cause
                is Error -> throw cause
                else -> throw RuntimeException("Unable to create binding instance.", cause)
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun findBindingConstructor(cls: Class<*>, clsSuffix: String, paramTypes: Array<Class<*>>):
        Constructor<BundleBinding>? {
        if (BINDINGS == null) {
            BINDINGS = WeakHashMap()
        }
        val bindingKey = cls.canonicalName + clsSuffix + paramTypes[0].simpleName
        var binding = BINDINGS!![bindingKey]
        if (binding != null) {
            debug("HIT: Cache found in binding weak map.")
            return binding
        }
        val clsName = cls.simpleName
        if (clsName.startsWith("android.") || clsName.startsWith("java.")) {
            debug("MISS: Reached framework class. Abandoning search.")
            return null
        }
        try {
            binding = cls.classLoader!!
                .loadClass(clsName + clsSuffix)
                .getConstructor(*paramTypes) as Constructor<BundleBinding>
            debug("HIT: Loaded binding class, caching in weak map.")
        } catch (e: ClassNotFoundException) {
            val supercls = cls.superclass!!
            debug("Not found. Trying superclass ${supercls.simpleName}")
            // abstract classes binding fix
            if (paramTypes[0] == cls && paramTypes[0].superclass == supercls) {
                paramTypes[0] = supercls
            }
            binding = findBindingConstructor(supercls, clsSuffix, paramTypes)
        } catch (e: NoSuchMethodException) {
            throw RuntimeException("Unable to find binding constructor for $clsName", e)
        }
        BINDINGS!![bindingKey] = checkNotNull(binding) {
            "Unable to find preferences binding, is `bundler-compiler` correctly installed?"
        }
        return binding
    }
}
